File: src/bin/swap_probe.rs
//! Verifies the swap.rs instruction builders against mainnet by assembling a
//! real swap and simulating it. We don't need funds: Anchor validates the
//! account context (PDAs, owners, tick arrays, oracle) BEFORE the handler
//! runs, so a correct build fails late at the token transfer (unfunded ATA),
//! while a wrong meta fails early with a constraint/seeds/owner error. The
//! probe prints the error class + logs so we can tell which happened.
//!
//! Usage: RPC_ENDPOINT=<url> cargo run --release --bin swap_probe

use arb_engine::execute::{
    decode_orca_state, decode_ray_state, orca_start_index, orca_tick_array, ray_start_index,
    ray_tick_array,
};
use arb_engine::pools::pair;
use arb_engine::swap::{
    orca_oracle, orca_swap_ix, ray_swap_ix, sqrt_limit, OrcaSwapAccounts, RaySwapAccounts,
};
use base64::Engine;
use solana_hash::Hash;
use solana_instruction::Instruction;
use solana_message::{legacy::Message, VersionedMessage};
use solana_pubkey::Pubkey;
use solana_signature::Signature;
use solana_transaction::versioned::VersionedTransaction;
use std::str::FromStr;

const ATA_PROGRAM: &str = "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL";
const TOKEN_PROGRAM: &str = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(resp) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = resp.into_json::<serde_json::Value>() {
                return Some(v);
            }
        }
        std::thread::sleep(std::time::Duration::from_millis(400 << attempt));
    }
    None
}

fn account_data(endpoint: &str, addr: &str) -> Option<Vec<u8>> {
    let v = rpc(
        endpoint,
        serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getAccountInfo",
            "params":[addr,{"encoding":"base64"}]}),
    )?;
    base64::engine::general_purpose::STANDARD
        .decode(v["result"]["value"]["data"][0].as_str()?)
        .ok()
}

fn pk_at(d: &[u8], o: usize) -> Pubkey {
    Pubkey::try_from(&d[o..o + 32]).unwrap()
}

fn ata(owner: &Pubkey, mint: &Pubkey) -> Pubkey {
    Pubkey::find_program_address(
        &[
            owner.as_ref(),
            Pubkey::from_str(TOKEN_PROGRAM).unwrap().as_ref(),
            mint.as_ref(),
        ],
        &Pubkey::from_str(ATA_PROGRAM).unwrap(),
    )
    .0
}

fn simulate(endpoint: &str, ix: Instruction, authority: Pubkey) -> (Option<String>, Vec<String>) {
    let msg = Message::new_with_blockhash(&[ix], Some(&authority), &Hash::default());
    let tx = VersionedTransaction {
        signatures: vec![Signature::default()],
        message: VersionedMessage::Legacy(msg),
    };
    let b64 = base64::engine::general_purpose::STANDARD.encode(bincode::serialize(&tx).unwrap());
    let v = rpc(
        endpoint,
        serde_json::json!({"jsonrpc":"2.0","id":1,"method":"simulateTransaction",
            "params":[b64,{"encoding":"base64","sigVerify":false,"replaceRecentBlockhash":true}]}),
    );
    let val = v.map(|v| v["result"]["value"].clone()).unwrap_or_default();
    let err = if val["err"].is_null() {
        None
    } else {
        Some(val["err"].to_string())
    };
    let logs = val["logs"]
        .as_array()
        .map(|a| a.iter().filter_map(|l| l.as_str().map(String::from)).collect())
        .unwrap_or_default();
    (err, logs)
}

fn classify(err: &Option<String>, logs: &[String]) -> &'static str {
    match err {
        None => "✅ SIMULATED OK — metas correct, swap executed",
        Some(_) => {
            let joined = logs.join("\n").to_lowercase();
            // Failing on OUR token accounts (unfunded/uninitialized ATAs) means
            // every pool-side meta already validated — the builder is correct.
            if joined.contains("insufficient")
                || joined.contains("input_token_account")
                || joined.contains("output_token_account")
                || joined.contains("token_owner_account")
                || joined.contains("3012")
                || joined.contains("not be already initialized")
                || joined.contains("could not create program address")
                || joined.contains("account not found")
                || joined.contains("uninitialized")
            {
                "✅ METAS OK — reached swap handler; only our unfunded ATA failed"
            } else if joined.contains("seeds")
                || joined.contains("constraint")
                || joined.contains("owned by")
                || joined.contains("declared program")
            {
                "❌ METAS WRONG — account-context validation failed"
            } else {
                "⚠️  INCONCLUSIVE — inspect logs below"
            }
        }
    }
}

fn main() {
    let _ = dotenvy::dotenv();
    let endpoint = std::env::var("RPC_ENDPOINT")
        .unwrap_or_else(|_| "https://api.mainnet-beta.solana.com".to_string());
    let cfg = pair();
    // Any wallet works — we're checking account structure, not balances.
    let authority = Pubkey::from_str("Anu6Awu4kxaEDrg1nkpcikx6tJ2xhfVci5TvDrZBsZEB").unwrap();

    // ── Orca ──
    println!("=== Orca {} pool {} ===", cfg.label, cfg.orca_pool);
    let od = account_data(&endpoint, &cfg.orca_pool).expect("orca pool");
    let ost = decode_orca_state(&od).expect("orca state");
    let orca_pk = Pubkey::from_str(&cfg.orca_pool).unwrap();
    let mint_a = pk_at(&od, 101);
    let mint_b = pk_at(&od, 181);
    let start = orca_start_index(ost.tick, ost.tick_spacing);
    // Three consecutive tick arrays in the swap direction (a_to_b = price down).
    let n = 88 * ost.tick_spacing as i32;
    let base_is_a = mint_a == Pubkey::from_str(&cfg.base_mint).unwrap();
    let a_to_b = base_is_a; // sell base = A→B when base is mintA
    let starts = if a_to_b {
        [start, start - n, start - 2 * n]
    } else {
        [start, start + n, start + 2 * n]
    };
    let oa = OrcaSwapAccounts {
        whirlpool: orca_pk,
        token_authority: authority,
        token_owner_a: ata(&authority, &mint_a),
        token_vault_a: pk_at(&od, 133),
        token_owner_b: ata(&authority, &mint_b),
        token_vault_b: pk_at(&od, 213),
        tick_arrays: [
            orca_tick_array(&orca_pk, starts[0]),
            orca_tick_array(&orca_pk, starts[1]),
            orca_tick_array(&orca_pk, starts[2]),
        ],
        oracle: orca_oracle(&orca_pk),
    };
    let ix = orca_swap_ix(&oa, 100_000, 0, sqrt_limit(a_to_b), a_to_b);
    let (err, logs) = simulate(&endpoint, ix, authority);
    println!("  a_to_b={a_to_b} err={err:?}");
    println!("  {}", classify(&err, &logs));
    for l in logs.iter().take(14) {
        println!("    {l}");
    }

    // ── Raydium CLMM ──
    println!("\n=== Raydium CLMM {} pool {} ===", cfg.label, cfg.ray_pool);
    let rd = account_data(&endpoint, &cfg.ray_pool).expect("ray pool");
    let rst = decode_ray_state(&rd).expect("ray state");
    let ray_pk = Pubkey::from_str(&cfg.ray_pool).unwrap();
    let amm_config = pk_at(&rd, 9);
    let mint0 = pk_at(&rd, 73);
    let mint1 = pk_at(&rd, 105);
    let vault0 = pk_at(&rd, 137);
    let vault1 = pk_at(&rd, 169);
    let observation = pk_at(&rd, 201);
    let base_is_0 = mint0 == Pubkey::from_str(&cfg.base_mint).unwrap();
    // Sell base: input is base. If base is mint0, input vault = vault0.
    let (input_mint, input_vault, output_vault) = if base_is_0 {
        (mint0, vault0, vault1)
    } else {
        (mint1, vault1, vault0)
    };
    let output_mint = if base_is_0 { mint1 } else { mint0 };
    let rstart = ray_start_index(rst.tick, rst.tick_spacing);
    let ra = RaySwapAccounts {
        payer: authority,
        amm_config,
        pool_state: ray_pk,
        input_token_account: ata(&authority, &input_mint),
        output_token_account: ata(&authority, &output_mint),
        input_vault,
        output_vault,
        observation_state: observation,
        tick_array: ray_tick_array(&ray_pk, rstart),
    };
    let is_base_input = true;
    let ix = ray_swap_ix(&ra, 100_000, 0, sqrt_limit(base_is_0), is_base_input);
    let (err, logs) = simulate(&endpoint, ix, authority);
    println!("  is_base_input={is_base_input} err={err:?}");
    println!("  {}", classify(&err, &logs));
    for l in logs.iter().take(14) {
        println!("    {l}");
    }
}

File: src/bin/swapProbe.ts
// Port of src/bin/swap_probe.rs
//
// Verifies the swap.rs instruction builders against mainnet by assembling a
// real swap and simulating it. We don't need funds: Anchor validates the
// account context (PDAs, owners, tick arrays, oracle) BEFORE the handler
// runs, so a correct build fails late at the token transfer (unfunded ATA),
// while a wrong meta fails early with a constraint/seeds/owner error. The
// probe prints the error class + logs so we can tell which happened.
//
// Usage: RPC_ENDPOINT=<url> tsx src/bin/swapProbe.ts

import 'dotenv/config';
import { PublicKey, TransactionInstruction, TransactionMessage, VersionedTransaction } from '@solana/web3.js';
import { decodeOrcaState, decodeRayState, orcaStartIndex, orcaTickArray, rayStartIndex, rayTickArray } from '../lib/execute.js';
import { pair } from '../lib/pools.js';
import { orcaOracle, orcaSwapIx, raySwapIx, sqrtLimit, type OrcaSwapAccounts, type RaySwapAccounts } from '../lib/swap.js';

const ATA_PROGRAM = 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL';
const TOKEN_PROGRAM = 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await res.json();
    } catch {
      // fall through to retry
    }
    await sleep(400 << attempt);
  }
  return undefined;
}

async function accountData(endpoint: string, addr: string): Promise<Uint8Array | undefined> {
  const v = await rpc(endpoint, { jsonrpc: '2.0', id: 1, method: 'getAccountInfo', params: [addr, { encoding: 'base64' }] });
  const b64 = v?.result?.value?.data?.[0];
  if (typeof b64 !== 'string') return undefined;
  return Uint8Array.from(Buffer.from(b64, 'base64'));
}

function pkAt(d: Uint8Array, o: number): PublicKey {
  return new PublicKey(d.subarray(o, o + 32));
}

function ata(owner: PublicKey, mint: PublicKey): PublicKey {
  return PublicKey.findProgramAddressSync([owner.toBuffer(), new PublicKey(TOKEN_PROGRAM).toBuffer(), mint.toBuffer()], new PublicKey(ATA_PROGRAM))[0];
}

async function simulate(endpoint: string, ix: TransactionInstruction, authority: PublicKey): Promise<[string | undefined, string[]]> {
  const msg = new TransactionMessage({
    payerKey: authority,
    recentBlockhash: '11111111111111111111111111111111',
    instructions: [ix],
  }).compileToLegacyMessage();
  const tx = new VersionedTransaction(msg);
  const b64 = Buffer.from(tx.serialize()).toString('base64');
  const v = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'simulateTransaction',
    params: [b64, { encoding: 'base64', sigVerify: false, replaceRecentBlockhash: true }],
  });
  const val = v?.result?.value ?? {};
  const err = val?.err === null || val?.err === undefined ? undefined : JSON.stringify(val.err);
  const logs: string[] = Array.isArray(val?.logs) ? val.logs.filter((l: unknown): l is string => typeof l === 'string') : [];
  return [err, logs];
}

function classify(err: string | undefined, logs: string[]): string {
  if (err === undefined) {
    return '✅ SIMULATED OK — metas correct, swap executed';
  }
  const joined = logs.join('\n').toLowerCase();
  // Failing on OUR token accounts (unfunded/uninitialized ATAs) means
  // every pool-side meta already validated — the builder is correct.
  if (
    joined.includes('insufficient') ||
    joined.includes('input_token_account') ||
    joined.includes('output_token_account') ||
    joined.includes('token_owner_account') ||
    joined.includes('3012') ||
    joined.includes('not be already initialized') ||
    joined.includes('could not create program address') ||
    joined.includes('account not found') ||
    joined.includes('uninitialized')
  ) {
    return '✅ METAS OK — reached swap handler; only our unfunded ATA failed';
  } else if (joined.includes('seeds') || joined.includes('constraint') || joined.includes('owned by') || joined.includes('declared program')) {
    return '❌ METAS WRONG — account-context validation failed';
  } else {
    return '⚠️  INCONCLUSIVE — inspect logs below';
  }
}

async function main(): Promise<void> {
  const endpoint = process.env.RPC_ENDPOINT ?? 'https://api.mainnet-beta.solana.com';
  const cfg = pair();
  // Any wallet works — we're checking account structure, not balances.
  const authority = new PublicKey('Anu6Awu4kxaEDrg1nkpcikx6tJ2xhfVci5TvDrZBsZEB');

  // ── Orca ──
  console.log(`=== Orca ${cfg.label} pool ${cfg.orcaPool} ===`);
  const od = await accountData(endpoint, cfg.orcaPool);
  if (od === undefined) throw new Error('orca pool');
  const ost = decodeOrcaState(od);
  if (ost === undefined) throw new Error('orca state');
  const orcaPk = new PublicKey(cfg.orcaPool);
  const mintA = pkAt(od, 101);
  const mintB = pkAt(od, 181);
  const start = orcaStartIndex(ost.tick, ost.tickSpacing);
  // Three consecutive tick arrays in the swap direction (a_to_b = price down).
  const n = 88 * ost.tickSpacing;
  const baseIsA = mintA.equals(new PublicKey(cfg.baseMint));
  const aToB = baseIsA; // sell base = A→B when base is mintA
  const starts = aToB ? [start, start - n, start - 2 * n] : [start, start + n, start + 2 * n];
  const oa: OrcaSwapAccounts = {
    whirlpool: orcaPk,
    tokenAuthority: authority,
    tokenOwnerA: ata(authority, mintA),
    tokenVaultA: pkAt(od, 133),
    tokenOwnerB: ata(authority, mintB),
    tokenVaultB: pkAt(od, 213),
    tickArrays: [orcaTickArray(orcaPk, starts[0]!), orcaTickArray(orcaPk, starts[1]!), orcaTickArray(orcaPk, starts[2]!)],
    oracle: orcaOracle(orcaPk),
  };
  const ix = orcaSwapIx(oa, 100_000n, 0n, sqrtLimit(aToB), true, aToB);
  const [err, logs] = await simulate(endpoint, ix, authority);
  console.log(`  a_to_b=${aToB} err=${err}`);
  console.log(`  ${classify(err, logs)}`);
  for (const l of logs.slice(0, 14)) {
    console.log(`    ${l}`);
  }

  // ── Raydium CLMM ──
  console.log(`\n=== Raydium CLMM ${cfg.label} pool ${cfg.rayPool} ===`);
  const rd = await accountData(endpoint, cfg.rayPool);
  if (rd === undefined) throw new Error('ray pool');
  const rst = decodeRayState(rd);
  if (rst === undefined) throw new Error('ray state');
  const rayPk = new PublicKey(cfg.rayPool);
  const ammConfig = pkAt(rd, 9);
  const mint0 = pkAt(rd, 73);
  const mint1 = pkAt(rd, 105);
  const vault0 = pkAt(rd, 137);
  const vault1 = pkAt(rd, 169);
  const observation = pkAt(rd, 201);
  const baseIs0 = mint0.equals(new PublicKey(cfg.baseMint));
  // Sell base: input is base. If base is mint0, input vault = vault0.
  const [inputMint, inputVault, outputVault] = baseIs0 ? [mint0, vault0, vault1] : [mint1, vault1, vault0];
  const outputMint = baseIs0 ? mint1 : mint0;
  // Selling base: input == base. zero_for_one when input is mint0 → arrays descend.
  const zeroForOne = baseIs0;
  const rn = 60 * rst.tickSpacing;
  const rstart = rayStartIndex(rst.tick, rst.tickSpacing);
  const rstarts = zeroForOne ? [rstart, rstart - rn, rstart - 2 * rn] : [rstart, rstart + rn, rstart + 2 * rn];
  const ra: RaySwapAccounts = {
    payer: authority,
    ammConfig,
    poolState: rayPk,
    inputTokenAccount: ata(authority, inputMint),
    outputTokenAccount: ata(authority, outputMint),
    inputVault,
    outputVault,
    observationState: observation,
    tickArrays: [rayTickArray(rayPk, rstarts[0]!), rayTickArray(rayPk, rstarts[1]!), rayTickArray(rayPk, rstarts[2]!)],
  };
  const isBaseInput = true;
  const rix = raySwapIx(ra, 100_000n, 0n, sqrtLimit(baseIs0), isBaseInput);
  const [rerr, rlogs] = await simulate(endpoint, rix, authority);
  console.log(`  is_base_input=${isBaseInput} err=${rerr}`);
  console.log(`  ${classify(rerr, rlogs)}`);
  for (const l of rlogs.slice(0, 14)) {
    console.log(`    ${l}`);
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
